File: app/src/main/java/com/fengwei23/androidsample/MainActivity.java
package com.fengwei23.androidsample;

import android.Manifest;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;
import com.sunplus.easypermissions.AfterPermissionGranted;
import com.sunplus.easypermissions.AppSettingsDialog;
import com.sunplus.easypermissions.EasyPermissions;
import java.util.Arrays;
import java.util.List;

public class MainActivity extends AppCompatActivity implements EasyPermissions.PermissionCallbacks,
    EasyPermissions.RationaleCallbacks {

  private static final String TAG = "MainActivity";
  private static final String[] LOCATION = {
      Manifest.permission.READ_CONTACTS,
      Manifest.permission.BLUETOOTH,
      Manifest.permission.BLUETOOTH_ADMIN,
      Manifest.permission.ACCESS_FINE_LOCATION,
      Manifest.permission.WRITE_EXTERNAL_STORAGE
  };
  private static final int RC_CAMERA_PERM = 123;
  private static final int RC_LOCATION_CONTACTS_PERM = 124;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    findViewById(R.id.button_camera).setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        cameraTask();
      }
    });

    findViewById(R.id.button_location_and_contacts).setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        locationAndContactsTask();
      }
    });
  }

  @AfterPermissionGranted(RC_LOCATION_CONTACTS_PERM)
  private void locationAndContactsTask() {
    if (hasLocationAndContactsPermissions()) {
      // Have permissions, do the thing!
      Toast.makeText(this, "TODO: Location and Contacts things", Toast.LENGTH_LONG).show();
    } else {
      // Ask for both permissions
      EasyPermissions.requestPermissions(
          this,
          getString(R.string.rationale_location_contacts),
          RC_LOCATION_CONTACTS_PERM,
          LOCATION);
    }
  }

  @AfterPermissionGranted(RC_CAMERA_PERM)
  private void cameraTask() {
    if (hasCameraPermission()) {
      Toast.makeText(this, "TODO: Camera things", Toast.LENGTH_LONG).show();
    } else {
      Toast.makeText(this, "NOT TODO: Camera things", Toast.LENGTH_LONG).show();
      EasyPermissions.requestPermissions(this,
          getString(R.string.rationale_camera),
          RC_CAMERA_PERM,
          Manifest.permission.CAMERA);
    }
  }

  private boolean hasCameraPermission() {
    return EasyPermissions.hasPermissions(this, Manifest.permission.CAMERA);
  }

  private boolean hasLocationAndContactsPermissions() {
    return EasyPermissions.hasPermissions(this, LOCATION);
  }

  private boolean hasStoragePermission() {
    return EasyPermissions.hasPermissions(this, Manifest.permission.WRITE_EXTERNAL_STORAGE);
  }

  @Override
  public void onPermissionsGranted(int requestCode, @NonNull List<String> perms) {
    Log.d(TAG, "onPermissionsGranted:" + requestCode + ":" + perms.size());
  }

  @Override
  public void onPermissionsDenied(int requestCode, @NonNull List<String> perms) {
    Log.d(TAG, "onPermissionsDenied:" + requestCode + ":" + perms.size());
    if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
      new AppSettingsDialog.Builder(this).build().show();
    }
  }

  @Override
  public void onRationaleAccepted(int requestCode) {
    Log.d(TAG, "onRationaleAccepted:" + requestCode);
  }

  @Override
  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                         @NonNull int[] grantResults) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    Log.d(TAG, "onRequestPermissionsResult: " + requestCode + ":" + Arrays.toString(grantResults));
  }

  @Override
  public void onRationaleDenied(int requestCode) {
    Log.d(TAG, "onRationaleDenied:" + requestCode);
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == AppSettingsDialog.DEFAULT_SETTINGS_REQ_CODE) {
      String yes = getString(R.string.yes);
      String no = getString(R.string.no);

      // Do something after user returned from app settings screen, like showing a Toast.
      Toast.makeText(
          this,
          getString(R.string.returned_from_app_settings_to_activity,
              hasCameraPermission() ? yes : no,
              hasLocationAndContactsPermissions() ? yes : no),
          Toast.LENGTH_LONG)
          .show();
    }
  }
}

File: screenrecorderdemo/src/main/java/com/almosting/sample/MainActivity.kt
package com.almosting.sample

import android.Manifest.permission
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.almosting.easypermissions.AfterPermissionGranted
import com.almosting.easypermissions.AppSettingsDialog
import com.almosting.easypermissions.AppSettingsDialog.Builder
import com.almosting.easypermissions.EasyPermissions
import com.almosting.easypermissions.EasyPermissions.PermissionCallbacks
import com.almosting.easypermissions.EasyPermissions.RationaleCallbacks
import com.almosting.sample.R.string
import com.almosting.sample.databinding.ActivityMainBinding
import java.util.Arrays

class MainActivity : AppCompatActivity(), PermissionCallbacks, RationaleCallbacks {
  private lateinit var binding: ActivityMainBinding
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = ActivityMainBinding.inflate(layoutInflater)
    setContentView(binding.root)
    binding.buttonCamera.setOnClickListener { cameraTask() }
    binding.buttonLocationAndContacts.setOnClickListener { locationAndContactsTask() }
  }

  @AfterPermissionGranted(RC_LOCATION_CONTACTS_PERM)
  private fun locationAndContactsTask() {
    if (hasLocationAndContactsPermissions()) {
      // Have permissions, do the thing!
      Toast.makeText(
        this,
        "TODO: Location and Contacts things",
        Toast.LENGTH_LONG
      ).show()
    } else {
      // Ask for both permissions
      EasyPermissions.requestPermissions(
        this,
        getString(string.rationale_location_contacts),
        RC_LOCATION_CONTACTS_PERM,
        LOCATION
      )
    }
  }

  @AfterPermissionGranted(RC_CAMERA_PERM)
  private fun cameraTask() {
    if (hasCameraPermission()) {
      Toast.makeText(this, "TODO: Camera things", Toast.LENGTH_LONG).show()
    } else {
      Toast.makeText(
        this,
        "NOT TODO: Camera things",
        Toast.LENGTH_LONG
      ).show()
      EasyPermissions.requestPermissions(
        this,
        getString(string.rationale_camera),
        RC_CAMERA_PERM,
        arrayOf(permission.CAMERA)
      )
    }
  }

  private fun hasCameraPermission(): Boolean {
    return EasyPermissions.hasPermissions(this, arrayOf(permission.CAMERA))
  }

  private fun hasLocationAndContactsPermissions(): Boolean {
    return EasyPermissions.hasPermissions(this, LOCATION)
  }

  private fun hasStoragePermission(): Boolean {
    return EasyPermissions.hasPermissions(this, arrayOf(permission.WRITE_EXTERNAL_STORAGE))
  }

  override fun onPermissionsGranted(
    requestCode: Int,
    perms: List<String>
  ) {
    Log.d(TAG, "onPermissionsGranted:" + requestCode + ":" + perms.size)
  }

  override fun onPermissionsDenied(
    requestCode: Int,
    perms: List<String>
  ) {
    Log.d(TAG, "onPermissionsDenied:" + requestCode + ":" + perms.size)
    if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
      Builder(this).build().show()
    }
  }

  override fun onRationaleAccepted(requestCode: Int) {
    Log.d(TAG, "onRationaleAccepted:$requestCode")
  }


  override fun onRequestPermissionsResult(
    requestCode: Int, permissions: Array<String>,
    grantResults: IntArray
  ) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    Log.d(TAG, "onRequestPermissionsResult: " + requestCode + ":" + Arrays.toString(grantResults))
  }

  override fun onRationaleDenied(requestCode: Int) {
    Log.d(TAG, "onRationaleDenied:$requestCode")
  }

  override fun onActivityResult(
    requestCode: Int,
    resultCode: Int,
    data: Intent?
  ) {
    super.onActivityResult(requestCode, resultCode, data)
    if (requestCode == AppSettingsDialog.DEFAULT_SETTINGS_REQ_CODE) {
      val yes = getString(string.yes)
      val no = getString(string.no)

      // Do something after user returned from app settings screen, like showing a Toast.
      Toast.makeText(
        this,
        getString(
          string.returned_from_app_settings_to_activity,
          if (hasCameraPermission()) yes else no,
          if (hasLocationAndContactsPermissions()) yes else no
        ),
        Toast.LENGTH_LONG
      ).show()
    }
  }

  companion object {
    private const val TAG = "MainActivity"
    private val LOCATION = arrayOf(
      permission.READ_CONTACTS,
      permission.BLUETOOTH,
      permission.BLUETOOTH_ADMIN,
      permission.ACCESS_FINE_LOCATION,
      permission.WRITE_EXTERNAL_STORAGE
    )
    private const val RC_CAMERA_PERM = 123
    private const val RC_LOCATION_CONTACTS_PERM = 124
  }
}